### ISSUE TITLE
linter: no-inner-declarations's `blockScopedFunctions:'allow'` default is only applied when passing an options object

### ISSUE THREAD

**Issue by BPScott** [oxc-project/oxc#25073]
### What version of Oxlint are you using?

1.75.0

### What does your `.oxlintrc.json` (or `oxlint.config.ts`) config file look like?

Raises Error:
```
{"rules": {"no-inner-declarations":"error"}}
```

No error:
```
{"rules": "no-inner-declarations": ["error", "functions", {"blockScopedFunctions": "allow"}]}
```

### What happened?

`blockScopedFunctions` is documented (and derives `#[default]`) as `allow`, but
enabling the rule with a bare severity — `"no-inner-declarations": "error"`, or
`-D no-inner-declarations` — behaves as `disallow`. 

Passing _any_ options array restores the documented default, even an array whose options object contains only an unrecognized key.

## Reproduction

The following code should pass linting - as it is a module file, which allows block scoped variables.

```js
export {};

{
  function inner() {}
  inner();
}
```

```console
$ oxlint -D no-inner-declarations index.js
index.js:4:3: error eslint(no-inner-declarations): Variable or `function` declarations are not allowed in nested blocks
  help: Move function declaration to program root
```

### Expected

No error. The following two configs should behave identically, as `"blockScopedFunctions": "allow"` is claimed to be the default behaviour.

```
{"rules": {"no-inner-declarations":"error"}}

{"rules": "no-inner-declarations": ["error", "functions", {"blockScopedFunctions": "allow"}]}
```

<details>
<summary>Claude generated investigation into the error cause. AI slop and I've got no Rust experience but this seems like it might be a reasonable starting point for a fix</summary>

## Cause

`crates/oxc_linter/src/rules/eslint/no_inner_declarations.rs`

The option is stored as an `Option`:

```rust
struct NoInnerDeclarationsOptions {
    #[schemars(with = "BlockScopedFunctions")]
    block_scoped_functions: Option<BlockScopedFunctions>,
    ...
}
```

`from_configuration` only populates it when the config value is an array, and then
falls back to `Allow` via `.or(...)`:

```rust
let block_scoped_functions = if value.is_array() && !value.is_null() {
    value
        .get(1)
        .and_then(|v| v.get("blockScopedFunctions"))
        .and_then(serde_json::Value::as_str)
        .map(|value| match value {
            "disallow" => BlockScopedFunctions::Disallow,
            _ => BlockScopedFunctions::Allow,
        })
        .or(Some(BlockScopedFunctions::Allow))
} else {
    None
};
```

and `run` gates the strict-mode exemption on an exact `Some(...)` match:

```rust
if self.0 == NoInnerDeclarationsConfig::Functions
    && self.1.block_scoped_functions == Some(BlockScopedFunctions::Allow)
{
    if ctx.source_type().is_module() {
        return;
    }
```

So with a bare severity the field is `None`, `None != Some(Allow)`, and the
exemption is skipped. The `#[default]` on `BlockScopedFunctions::Allow` is never
consulted, because nothing ever calls `unwrap_or_default()` on the field.

## Two related observations

1. **`namespaces` has the same shape and escapes by luck.** It is also
   `Option<Namespaces>` and also compared as `self.1.namespaces == Some(Namespaces::Allow)`,
   but its derived default is `Disallow` — so the `None` case happens to coincide
   with the intended default. The same `Option`-compared-to-`Some` pattern in other
   rules would be worth auditing; it is only observable where the derived default
   is the non-`None`-equivalent variant.

2. **`deny_unknown_fields` on `NoInnerDeclarationsOptions` is dead.**
   `from_configuration` hand-parses `value.get(1)` rather than deserializing the
   struct, so `{"notARealOption": "x"}` is silently accepted instead of erroring.
   The derived `JsonSchema` also advertises `blockScopedFunctions` as a plain
   (non-optional) enum defaulting to `allow`, so schema-driven editor tooling and
   the docs both disagree with runtime behaviour.

## Suggested fix

Any of:

- drop the `Option` wrappers and rely on `#[serde(default)]` for the derived defaults;
- keep the `Option` but compare with `self.1.block_scoped_functions.unwrap_or_default() == BlockScopedFunctions::Allow` in `run`;
- or have `from_configuration` end with `unwrap_or_default()` instead of the
  `is_array()`-guarded `.or(Some(Allow))`.

The first also lets the options struct be deserialized properly, which would make
`deny_unknown_fields` effective again.

</details>
